lib/spack/spack/solver/asp.py

```diff
@@ -1734,13 +1734,32 @@ def key_fn(item):
             # All the preferred version from packages.yaml, versions in external
             # specs will be computed later
             version_preferences = packages_yaml.get(pkg_name, {}).get("version", [])
-            for idx, v in enumerate(version_preferences):
+            idx = 0
+            for v in version_preferences:
                 # v can be a string so force it into an actual version for comparisons
-                ver = vn.Version(v)
-                self.declared_versions[pkg_name].append(
-                    DeclaredVersion(version=ver, idx=idx, origin=Provenance.PACKAGES_YAML)
-                )
-                self.possible_versions[pkg_name].add(ver)
+                pkg_class = spack.repo.path.get_pkg_class(pkg_name)
+                if vn.Version(v) in pkg_class.versions:
+                    version_defs = [vn.Version(v)]
+                else:
+                    ver = vn.ver(v)
```

Comment:
Currently on `develop`, the `versions` config only accepts a list of concrete versions, not ranges. Here you're interpreting a version first as a concrete one, and if that fails, as a range. That feels backwards to me... either parse with `vn.ver(v)` and use satisfies throughout, or stick to specific versions with `vn.Version(v)` like on develop, but not some hybrid that can only cause confusion.

Reply:
Let's just go for that option, `VersionList(version_preferences)` looks much cleaner to me.

Reply:
To be clear, you are open to the idea that users could write something like:

```
packages:
  python:
    version: ["=3.5"]
```

if they want exactly `3.5` and that otherwise (i.e. if they write `version: ["3.5"]`) then we could interpret it as a range. That seems reasonable to me (and aligns the version preference syntax with spec syntax).

Reply:
Yeah, it's a breaking change, but a very small breakage... "3.5" would match "3.5.1" etc